### ISSUE TITLE
Phoenix Presence Zombies when application is deployed.

### ISSUE THREAD

**Issue by alvesl** [phoenixframework/phoenix#2306]
We were able to verify (even locally) the presence of zombies in our network, and would like to share our findings and explore possible solutions since Presence is a key-factor on our application.

First, I'll copy and paste the ongoing discussion from a different issue (anyone interested, please see  https://github.com/phoenixframework/phoenix/issues/2255 which IS NOT related, but contains part of the discussion):

@dmorneau describing zombie behavior for him:
https://github.com/phoenixframework/phoenix/issues/2255#issuecomment-298308462

Our behavior: 
https://github.com/phoenixframework/phoenix/issues/2255#issuecomment-298475060

--- 
So, it turns out that after a lot of testing we could actually reduce the problem, although the solution might be very specific to our problem (we use a GenServer to actually startup and connect our Elixir nodes). Although it doesn't completely fix it, it sure helps reduce the number of zombies.
I'll briefly describe the problem first, then our fix and would welcome any new ideas on how to make this better to get us rid of zombies once and for all.

### Reproducing
We were able to reproduce the issue locally, which indicates that this is not related to running nodes on docker or anything. You should be able to reproduce this (we think) in any system that tracks presence and connects to multiple nodes.
1. First, get a project that properly tracks a user ID going, then start your first node:
`iex --sname n1 -S mix phoenix.server`
Now connect some clients to it (we like to connect 5 clients with the same userID-1 to reproduce this, so basically we get a few connections going, each one with a different ref)
2. Now run server two:
`iex --sname n2 -S mix phoenix.server`
Then connect 5 more clients with a different user, userID-2 to n2.
3. Now lets confirm each Presence list separately and make sure they both show 5 connections  from users ID 1 and 2.
4. Time to merge our servers, we do that by calling:
`Node.connectt :'n2@ip-host'`
5. Our presence list is now merged! Now if you query the presence in both servers it properly shows 10 connections, from 2 users.
6. Let's kill n2 (just exit the shell) and bring up a new server n3:
`iex --sname n3 -S mix phoenix.server`
7. If you now look into n1, you'll see that it properly updated its list when n2 went down. It now shows 5 users, *the problem is seen when you then connect n3 to n1*!!!
8. Call Node.connect again - it does not even matter if n3 has no clients connected to it, when it joins n1, the presence list is updated *bringing back zombies from our now dead n2 node* 😱 .

To make the problem even more clear, we added the node name to the presence list as we track our users, and we can clearly see that the zombies that come back on step 8 are actually from n2!!!
It almost feels like the nodes were NOT completely removed from the shared structure. I did not find out exactly how it works under the hood, but from DockYards blog post:
> With Phoenix, we've developed a system based on a distributed heartbeat/gossip protocol which uses a CRDT to replicate data in a conflict-free way

I can't help to think that something went wrong there when the two nodes are trying to determine who is right or wrong about the correct data.

After all this, we couldn't come up with a way to properly fix it, only minimize the problem. Our assumption is based on thinking that it happens because Elixir was not gracefully terminated (even though we think that it should auto-heal). So what we did is:
1. We changed the way we start our scripts  (docker files) to actually make sure they pass SIGTERM information to the underlying processes it starts. You can do this by changing something like:
`-CMD /opt/app/bin/boot.sh foreground`
to
`ENTRYPOINT ["/opt/app/bin/boot.sh", "foreground"]`

2. Then, in our GenServer (which boots and connects all our nodes) we trapped the exit signal:
`Process.flag(:trap_exit, true)` and tried to gracefully terminate:
```  
def terminate(reason, _state) do
  Logger.debug "#{inspect Node.self()} is going down: #{inspect reason}"
  Phoenix.Tracker.graceful_permdown(Project.Presence)
end
```
The reason why we think sometimes we still get ghosts is because once in a while, during a deploy, a timeout might be hit and `terminate` has not enough time to gracefully clean up and exit.

Any ideas? Anything we could improve on, or help you debug this? I'm sorry for not providing code for the actual testing, but let me know if you find it difficult to reproduce and we can do that.
Thanks again for all the help!

**Comment by alvesl** [comment]
@chrismccord any updates on this? I think it is a pretty big use-case problem with presence.
Anything I can do to help us move forward on the issue?
Thanks!

**Comment by chrismccord** [comment]
@alvesl Yes, @Gazler and I have been working to replicate this issue and spec out a fix. As of today, we have a repeatable test case which reproduces it and I know what's needed for the fix. Time willing, I should have something in place this week, but I have other priorities in play so don't hold me to it :)

**Comment by alvesl** [comment]
Oh awesome news, no problem at all.. I was just worried it would be hard to replicate by my description, please take your time & thanks for the amazing work!

**Comment by chrismccord** [comment]
@alvesl I believe I've patched this issue on the `cm-clouds` branch. If you have time to test it out on your app under the scenarios that caused your issues, that would be extremely appreciated. Thanks!

https://github.com/phoenixframework/phoenix_pubsub/pull/80

    {:phoenix_pubsub, github: "phoenixframework/phoenix", branch: "cm-clouds", override: true}

**Comment by alvesl** [comment]
Absolutely! Thank you so much for the fix, will run the tests and let you know.

**Comment by alvesl** [comment]
Good news @chrismccord @Gazler - I'm unable to reproduce ! 🎉
You guys are awesome! I can also now see the heartbeat being exchanged and the data exchange happening.
Here's a snapshot of my final structure (no n2 zombies!):

```
%{"1" => %{metas: [%{client: "web", node: :"n3@Lucass-MacBook-Pro",
       phx_ref: "0ugBAMOjKY0=", since: 1497403191, status: "online"},
     %{client: "web", node: :"n3@Lucass-MacBook-Pro", phx_ref: "aM4mcnpzsHw=",
       since: 1497403191, status: "online"},
     %{client: "web", node: :"n3@Lucass-MacBook-Pro", phx_ref: "uEYjHTyPTmY=",
       since: 1497403191, status: "online"},
     %{client: "web", node: :"n3@Lucass-MacBook-Pro", phx_ref: "SmHSbnmLKGM=",
       phx_ref_prev: "vsvY8ZIzCxs=", since: 1497403354, status: "away"},
     %{client: "web", node: :"n3@Lucass-MacBook-Pro", phx_ref: "Dt4DKHXGlc0=",
       since: 1497403196, status: "online"},
     %{client: "web", node: :"n1@Lucass-MacBook-Pro", phx_ref: "pFnm4Kfd4x0=",
       since: 1497401913, status: "online"},
     %{client: "web", node: :"n1@Lucass-MacBook-Pro", phx_ref: "E59N6GH2kvA=",
       since: 1497401937, status: "online"},
     %{client: "web", node: :"n1@Lucass-MacBook-Pro", phx_ref: "SoMWEFMY6ck=",
       since: 1497401938, status: "online"},
     %{client: "web", node: :"n1@Lucass-MacBook-Pro", phx_ref: "gUp3prKVmdc=",
       since: 1497401939, status: "online"},
     %{client: "web", node: :"n1@Lucass-MacBook-Pro", phx_ref: "qXlDoTGK7VM=",
       phx_ref_prev: "3r3KDm/fvHg=", since: 1497402727, status: "online"}]}}
```
Thanks for the  great work!

**Comment by chrismccord** [comment]
Fantastic! Thanks for following up! I should have a new version of phoenix_pubsub pushed soon.

**Comment by chrismccord** [comment]
Fixed in phoenix_pubsub 1.0.2
